add idle.finance custom integration

ajb opened the pull request.

ajb commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+        address[] memory outputTokens = new address[](1);
+        outputTokens[0] = IIdleToken(vault).token();
+        uint256[] memory outputAmounts = new uint256[](1);
+        outputAmounts[0] = IIdleToken(vault).tokenPriceWithFee(_strategy);
```
They have both `tokenPrice()` and `tokenPriceWithFee(address user)` methods, I am pretty sure we want the latter. Is `_strategy` the right address to pass?

rrecuero replied:
You still need to multiply this price by the amount of liquidity that is deposited right?

ajb replied:
Makes sense, thanks!

ajb commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+        bytes calldata _data,
+        address _tokenDenominator
+    ) external view override returns (uint256) {
+        address vault = BytesLib.decodeOpDataAddress(_data);
```
Copied from Yearn example, does this need to be here though?

rrecuero replied:
Yes, and you need to multiply this price by the Idle Token

Idle Token -> Amount of Underlying -> Amount of reserve asset

```
return IIdleToken(vault).tokenPrice().preciseMul(
_getPrice(IIdleToken(vault).token(), _tokenDenominator));
```

ajb commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+        address _vaultAddress
+    ) internal view returns (bool) {
+      // check for an Idle vault by seeing if it responds to the token() method with a token address
+      (bool success, bytes memory data) = address(_vaultAddress).staticcall{gas: 10000}(abi.encodeWithSignature('token()'));
```
Dunno what y'all think about this, or if you have a pattern for this type of stuff already? AFAIK there is no "vault registry" for Idle like there is for Yearn.

rrecuero replied:
I would use the idle price helper for this 

https://etherscan.io/address/0x04Ce60ed10F6D2CfF3AA015fc7b950D13c113be5

If it has a price, it is a valid idle token

ajb replied:
isn't this going to raise an exception if we pass an invalid token? i assumed that we are trying to avoid that?

ajb replied:
or wait, you're saying to keep the low level call, but call the price helper? 👌

ajb replied:
or solidity try...catch works too, i suppose?

rrecuero commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+     * hparam  _data                      Address provided as param
+     * @return address[] memory           List of reward token addresses
+     */
+    function _getRewardTokens(
```
you can delete this function. No need to override it if there is no change

rrecuero commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+     * @return uint256                   Call value
+     * @return bytes                     Trade calldata
+     */
+    function _getPreActionCallData(
```
you can delete this function if it is not needed

rrecuero commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    /**
+     * (OPTIONAL) Return post action calldata
```
delete this function as well

rrecuero commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+     * @return address                    Address of the asset to approve
+     * @return address                    Address to approve
+     */
+    function _preActionNeedsApproval(
```
delete

rrecuero commented on contracts/integrations/custom/CustomIntegrationIdleFinance.sol:
```diff
@@ -0,0 +1,338 @@
+     * @return address                    Address of the asset to approve
+     * @return address                    Address to approve
+     */
+    function _postActionNeedsApproval(
```
delete

rrecuero commented on test/fork/integration.test.js:
```diff
@@ -194,4 +194,52 @@ describe('Babylon integrations', function () {
     await customStrategy.connect(keeper).finalizeStrategy(0, '', 0);
   });
+
+  it('can deploy a strategy with the Idle.Finance Custom integration', async () => {
```
would be good to have two arrays with all the vaults. One for best yield and one for senior tranches and then have a loop for each and then run the tests for each vault.

bestYieldVaults.forEach((v) => {
  it('can enter into the Idle vault ${v}', () => {
   ...
  }); 
})

ajb replied:
This is done, however I'm a bit confused about how Idle works now.

I added the array for the "best yield" vaults and they all seem to pass the tests.

However it looks like the "Perpetual Yield Tranches" is a different product completely: https://docs.idle.finance/developers/perpetual-yield-tranches/architecture

Seems like maybe this needs to be a separate Babylon integration, since it's a completely different interface?

rrecuero replied:
This one looks great. 

Yes, I recommend doing a different integration. That integration would receive two params in bytes the address of the vault like the best yield plus a number. If the number is 0 then it is the junior tranche if the number is one it's the senior tranche. In the post-action call data for enter I would stake the tokens and in the post-action call data for exit I would unstake (it also claims).

Also take into account that for the convex ones, you would need several operations (curve->convex->idle tranche). See the yearn example to see how you can chain them.

Given that these two integrations are kind of the same but it's more work for you, wdyt about getting a 7.5k bounty in BABL for both?

ajb replied:
Sounds good! Shouldn't be an issue - just wanted to pick something simple to start. 

Btw, one last thing I realized - I can't quite tell if there are additional farming rewards for these "best yield" vaults. This makes me think that there are:

<img width="709" alt="Screen Shot 2022-05-28 at 3 59 18 PM" src="https://user-images.githubusercontent.com/1270317/170845302-b793a516-6d03-4bda-ac7e-1c813e40cec1.png">


I'm not sure exactly how this works, as I didn't see it in the interface. But it looks like upon calling `redeemIdleToken`, you actually receive various other tokens? See this tx for an example: <https://etherscan.io/tx/0x98311622aea3211302abc9ffe34559ba58735d36dc0b204d21ec54d37a9039f3>

Looks like I might need to update this PR.

rrecuero replied:
Yes, you need to implement this function:

```
    function _getRewardTokens(
        address /* _data */
    ) internal view virtual returns (address[] memory) {
        return new address[](1);
    }
```

Loop over `govTokens` and return them

ajb replied:
Should be all set now 👍

ajb replied:
> Given that these two integrations are kind of the same but it's more work for you, wdyt about getting a 7.5k bounty in BABL for both?

Missed this part - sounds good!

ajb commented:
Thanks @rrecuero, super helpful comments. Take a look to see if this addresses them :)

ajb commented:
Ok @rrecuero, cleaned everything up and squashed the commits. Tests passing for me:

<img width="1624" alt="Screen Shot 2022-05-29 at 7 38 18 PM" src="https://user-images.githubusercontent.com/1270317/170907192-cf409cce-2a1c-4bcc-905a-4bf0d759587c.png">

I think this is RFR again :)
